src/abstractblocksparsearray/abstractblocksparsematrix.jl

```diff
@@ -1 +1,76 @@
 const AbstractBlockSparseMatrix{T} = AbstractBlockSparseArray{T,2}
+
+# SVD is implemented by trying to
+# 1. Attempt to find a block-diagonal implementation by permuting
+# 2. Fallback to AbstractBlockArray implementation via BlockedArray
+
+function eigencopy_oftype(A::AbstractBlockSparseMatrix, T)
+  if is_block_permutation_matrix(A)
+    Acopy = similar(A, T)
+    for bI in eachblockstoredindex(A)
+      Acopy[bI] = eigencopy_oftype(A[bI], T)
+    end
+    return Acopy
+  else
+    return BlockedMatrix{T}(A)
+  end
+end
+
+function is_block_permutation_matrix(a::AbstractBlockSparseMatrix)
+  return allunique(first ∘ Tuple, eachblockstoredindex(a)) &&
+         allunique(last ∘ Tuple, eachblockstoredindex(a))
+end
+
+function _allocate_svd_output(A::AbstractBlockSparseMatrix, full::Bool, ::Algorithm)
+  @assert !full "TODO"
```

Comment:
What are you thinking for `full = true`? Do you think this function can reasonably be generalized or do you need a separate function?

Reply:
I am guessing it should be possible to generalize this. The main issue here is that the amount of ambiguity increases with the number of nonzero rows or columns, so there is a certain degree of arbitrariness in how the extra blocks would be filled in.

Reply:
Yeah it definitely gets more complicated in that regard. Though I guess in some ways it is fixed, since the axes should be something like:
```julia
  U = similar(A, axes(A, 1), axes(A, 1))
  S = similar(A, [axes(A, 1) or axes(A, 2) depending on which one is smaller])
  Vt = similar(A, axes(A, 2), axes(A, 2))
```

Reply:
The problem is mostly that it is now also possible for `axes(A,1)` to be both larger and smaller than `axes(A, 2)`, in the sense that you can have both thin and wide blocks. The overall structure should remain mostly the same I think, first pairing up the blocks, than taking the largest axis of each block, but it is a bit more annoying since `S` now has different axes from `U` and `V` if we keep to the LinearAlgebra convention where only `U` and `V` actually change

Reply:
Ah good point about rectangular blocks, that is definitely an extra layer of subtlety to account for.